File: services/web-test/app/api/bgg_sync.py
"""Proxy-роутер для UI BGG Sync.

Под `/api/bgg-sync/*` форвардим вызовы к catalog'у:
- история ImportJob с фильтрами (`/import/jobs`),
- управление scheduler (`/scheduler/jobs/...`),
- read-API снимков Hotness и GeekList (`/bgg/hotness/...`, `/bgg/geeklists/...`),
- запуск ручного импорта GeekList и mini-batch.

Архитектура — тонкий слой: каждый endpoint форвардит в `CatalogClient`,
маппит `CatalogServiceError` → HTTPException. Никакой бизнес-логики.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query

from app.catalog_client import CatalogClient, CatalogServiceError
from app.deps import get_catalog_client

router = APIRouter(prefix="/bgg-sync", tags=["bgg-sync"])


def _err(e: CatalogServiceError) -> HTTPException:
    return HTTPException(status_code=e.status_code, detail=e.detail)


# ── Scheduler control ─────────────────────────────────────────────────────────


@router.get("/scheduler/jobs")
async def list_scheduler_jobs(
    client: CatalogClient = Depends(get_catalog_client),
) -> list[dict]:
    """Список scheduler-job'ов с runtime info (next_run, last_run)."""
    try:
        return await client.list_scheduler_jobs()
    except CatalogServiceError as e:
        raise _err(e) from e


@router.patch("/scheduler/jobs/{job_id}")
async def reschedule_job(
    job_id: str,
    payload: dict,
    client: CatalogClient = Depends(get_catalog_client),
) -> dict:
    """Изменить cron/enabled/params + hot-reload в scheduler'е."""
    try:
        return await client.reschedule_job(job_id, payload)
    except CatalogServiceError as e:
        raise _err(e) from e


@router.post("/scheduler/jobs/{job_id}/trigger")
async def trigger_job(
    job_id: str,
    client: CatalogClient = Depends(get_catalog_client),
) -> dict:
    """Manual trigger scheduled-job'а — создаёт ImportJob с trigger='manual'."""
    try:
        return await client.trigger_scheduler_job(job_id)
    except CatalogServiceError as e:
        raise _err(e) from e


# WT-F7 bulk actions. job_id-сегменты `pause-all`/`resume-all`/`trigger-overdue`
# реально не job_id'ы, а action-имена — они НЕ пересекаются с разрешёнными
# job_id (snake_case в БД) благодаря тире. Если позднее придёт job_id вида
# `pause-all` — это поломает роутинг; пока маловероятно.
_BULK_ACTIONS = {"pause-all", "resume-all", "trigger-overdue"}


@router.post("/scheduler/jobs/{action}")
async def bulk_scheduler_action(
    action: str,
    client: CatalogClient = Depends(get_catalog_client),
) -> dict:
    """WT-F7: bulk action над всеми scheduler-job'ами."""
    if action not in _BULK_ACTIONS:
        raise HTTPException(status_code=404, detail=f"unknown bulk action: {action}")
    try:
        return await client.scheduler_bulk_action(action)
    except CatalogServiceError as e:
        raise _err(e) from e


@router.get("/settings/bgg")
async def get_bgg_settings(
    client: CatalogClient = Depends(get_catalog_client),
) -> dict:
    """WT-F7: сводка Global BGG Settings — token flag + cascade-настройки."""
    try:
        return await client.get_bgg_runtime_summary()
    except CatalogServiceError as e:
        raise _err(e) from e


# ── Job history ───────────────────────────────────────────────────────────────


@router.get("/jobs")
async def list_jobs(
    type: str | None = Query(None),
    status: str | None = Query(None),
    trigger: str | None = Query(None, description="manual | scheduled | api"),
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
    client: CatalogClient = Depends(get_catalog_client),
) -> list[dict]:
    """История ImportJob'ов с фильтрами для таблицы истории UI."""
    try:
        return await client.list_import_jobs(
            type=type,
            status=status,
            trigger=trigger,
            limit=limit,
            offset=offset,
        )
    except CatalogServiceError as e:
        raise _err(e) from e


# ── Manual imports ────────────────────────────────────────────────────────────


@router.post("/imports/geeklist")
async def import_geeklist(
    payload: dict,
    client: CatalogClient = Depends(get_catalog_client),
) -> dict:
    """Запуск snapshot'а кураторского BGG GeekList'а.

    payload: {geeklist_id: int, auto_import?: bool}.
    """
    try:
        return await client.import_bgg_geeklist(payload)
    except CatalogServiceError as e:
        raise _err(e) from e


@router.post("/imports/mini-batch")
async def import_mini_batch(
    payload: dict,
    client: CatalogClient = Depends(get_catalog_client),
) -> dict:
    """Daily catch-up enrichment (500-1000 игр)."""
    try:
        return await client.import_bgg_mini_batch(payload)
    except CatalogServiceError as e:
        raise _err(e) from e


# ── BGG snapshots (Hotness + GeekList) ────────────────────────────────────────


@router.get("/hotness/dates")
async def hotness_dates(
    limit: int = Query(30, ge=1, le=365),
    client: CatalogClient = Depends(get_catalog_client),
) -> list[str]:
    """Список snapshot_date hotness в БД (ISO YYYY-MM-DD), DESC."""
    try:
        return await client.bgg_hotness_dates(limit=limit)
    except CatalogServiceError as e:
        raise _err(e) from e


@router.get("/hotness")
async def hotness_snapshot(
    date: str | None = Query(None, description="YYYY-MM-DD"),
    client: CatalogClient = Depends(get_catalog_client),
) -> dict:
    """Snapshot hotness на дату (или последний доступный)."""
    try:
        return await client.bgg_hotness_snapshot(date)
    except CatalogServiceError as e:
        raise _err(e) from e


@router.get("/geeklists")
async def list_geeklists(
    client: CatalogClient = Depends(get_catalog_client),
) -> list[dict]:
    """Список импортированных GeekList'ов с последним snapshot_date."""
    try:
        return await client.bgg_geeklists()
    except CatalogServiceError as e:
        raise _err(e) from e


@router.get("/geeklists/{geeklist_id}")
async def get_geeklist(
    geeklist_id: int,
    date: str | None = Query(None),
    client: CatalogClient = Depends(get_catalog_client),
) -> dict:
    """Snapshot одного GeekList'а на дату."""
    try:
        return await client.bgg_geeklist_snapshot(
            geeklist_id, snapshot_date=date
        )
    except CatalogServiceError as e:
        raise _err(e) from e

File: services/web-test/frontend/src/lib/bgg-sync.ts
/**
 * Клиент к /api/bgg-sync/* — proxy на catalog'овые /scheduler/*, /import/jobs,
 * /import/bgg/{geeklist,mini-batch}, /bgg/{hotness,geeklists}.
 *
 * Типы зеркалят response_model из catalog/schemas.py:
 *   SchedulerJobOut, ImportJobOut.
 *
 * Cache-key namespace в react-query — `['bgg-sync', ...]`.
 */
import type { ImportJob } from './catalog'

const BASE = '/api/bgg-sync'

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) throw new Error(`${res.status} ${await res.text()}`)
  return res.json() as Promise<T>
}

// ── Scheduler ────────────────────────────────────────────────────────────────

/**
 * WT-F7: описание одного поля в schema-driven форме параметров job'а.
 * Зеркало catalog/scheduler.py:FieldSpec.
 */
export type FieldSpec = {
  name: string
  type: 'int' | 'float' | 'bool' | 'string' | 'enum'
  label: string
  description?: string
  default: unknown
  required?: boolean
  enum?: string[]
  min?: number
  max?: number
}

/** Один scheduler-job: конфиг из БД + runtime info из APScheduler + meta. */
export type SchedulerJob = {
  job_id: string
  cron_expr: string
  enabled: boolean
  params: Record<string, unknown>
  // Денормализация — обновляется при каждом trigger'е.
  last_run_job_id: number | null
  last_run_status: 'pending' | 'running' | 'done' | 'failed' | null
  last_run_at: string | null
  // Из APScheduler runtime.
  next_run_at: string | null
  // Из реестра JOB_METADATA.
  display_name: string | null
  description: string | null
  // Ring-buffer тиков для interval-job'ов (заполнен для match_worker/ml_health/etc.).
  tick_history?: Array<{ ts: string; duration_ms: number; error: boolean }>
  // WT-F7: если null — UI показывает JSON-textarea (legacy fallback).
  params_schema: FieldSpec[] | null
}

export type RescheduleRequest = {
  cron_expr?: string
  enabled?: boolean
  params?: Record<string, unknown>
}

/** WT-F7: ответ bulk-action endpoint'ов. */
export type SchedulerBulkActionResult = {
  action: string
  affected: string[]
  triggered_import_job_ids: number[]
  errors: Array<{ job_id: string; error: string }>
}

/** WT-F7: сводка Global BGG Settings. */
export type BggRuntimeSummary = {
  bgg_api_token_set: boolean
  family_cascade_enabled: boolean
  family_cascade_enabled_editable: boolean
  family_cascade_rate_limit_sec: number
}

export const fetchSchedulerJobs = () =>
  fetch(`${BASE}/scheduler/jobs`).then(r => json<SchedulerJob[]>(r))

export const rescheduleJob = (jobId: string, payload: RescheduleRequest) =>
  fetch(`${BASE}/scheduler/jobs/${encodeURIComponent(jobId)}`, {
    method: 'PATCH',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  }).then(r => json<SchedulerJob>(r))

export const triggerSchedulerJob = (jobId: string) =>
  fetch(`${BASE}/scheduler/jobs/${encodeURIComponent(jobId)}/trigger`, {
    method: 'POST',
  }).then(r => json<ImportJob>(r))

// WT-F7 bulk actions.
export const pauseAllJobs = () =>
  fetch(`${BASE}/scheduler/jobs/pause-all`, { method: 'POST' })
    .then(r => json<SchedulerBulkActionResult>(r))

export const resumeAllJobs = () =>
  fetch(`${BASE}/scheduler/jobs/resume-all`, { method: 'POST' })
    .then(r => json<SchedulerBulkActionResult>(r))

export const triggerOverdueJobs = () =>
  fetch(`${BASE}/scheduler/jobs/trigger-overdue`, { method: 'POST' })
    .then(r => json<SchedulerBulkActionResult>(r))

export const fetchBggSettings = () =>
  fetch(`${BASE}/settings/bgg`).then(r => json<BggRuntimeSummary>(r))

// ── Import history ───────────────────────────────────────────────────────────

export type ImportJobsFilter = {
  type?: string
  status?: 'pending' | 'running' | 'done' | 'failed'
  trigger?: 'manual' | 'scheduled' | 'api'
  limit?: number
  offset?: number
}

export const fetchImportJobs = (filter: ImportJobsFilter = {}) => {
  const qs = new URLSearchParams()
  if (filter.type) qs.set('type', filter.type)
  if (filter.status) qs.set('status', filter.status)
  if (filter.trigger) qs.set('trigger', filter.trigger)
  qs.set('limit', String(filter.limit ?? 50))
  qs.set('offset', String(filter.offset ?? 0))
  return fetch(`${BASE}/jobs?${qs}`).then(r => json<ImportJob[]>(r))
}

// ── Manual triggers ──────────────────────────────────────────────────────────

export type GeeklistImportPayload = {
  geeklist_id: number
  auto_import?: boolean
}

export const importBggGeeklist = (payload: GeeklistImportPayload) =>
  fetch(`${BASE}/imports/geeklist`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  }).then(r => json<ImportJob>(r))

export type MiniBatchImportPayload = {
  batch_size?: number
  skip_recent_days?: number
  rate_limit_sec?: number
  dry_run?: boolean
}

export const importBggMiniBatch = (payload: MiniBatchImportPayload = {}) =>
  fetch(`${BASE}/imports/mini-batch`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  }).then(r => json<ImportJob>(r))

// ── BGG snapshots (Hotness + GeekList) ───────────────────────────────────────

export type HotnessItem = {
  rank: number
  bgg_id: number
  name: string
  year: number | null
  thumbnail_url: string | null
  // Денормализация: catalog game_id если игра уже в каталоге.
  game_id: number | null
  game_title: string | null
}

export type HotnessSnapshot = {
  snapshot_date: string | null
  items: HotnessItem[]
}

export const fetchHotnessDates = (limit = 30) =>
  fetch(`${BASE}/hotness/dates?limit=${limit}`).then(r => json<string[]>(r))

export const fetchHotnessSnapshot = (date?: string) => {
  const url = date ? `${BASE}/hotness?date=${date}` : `${BASE}/hotness`
  return fetch(url).then(r => json<HotnessSnapshot>(r))
}

export type GeeklistMeta = {
  geeklist_id: number
  latest_snapshot_date: string
  title: string | null
  username: string | null
  item_count: number
}

export type GeeklistItem = {
  rank: number
  bgg_id: number
  name: string
  body: string | null
  game_id: number | null
  game_title: string | null
}

export type GeeklistSnapshot = {
  geeklist_id: number
  snapshot_date: string
  title: string | null
  description: string | null
  username: string | null
  item_count: number
  items: GeeklistItem[]
  fetched_at: string
}

export const fetchGeeklists = () =>
  fetch(`${BASE}/geeklists`).then(r => json<GeeklistMeta[]>(r))

export const fetchGeeklistSnapshot = (geeklistId: number, date?: string) => {
  const url = date
    ? `${BASE}/geeklists/${geeklistId}?date=${date}`
    : `${BASE}/geeklists/${geeklistId}`
  return fetch(url).then(r => json<GeeklistSnapshot>(r))
}
